---
title: "GGI Workshop demo maize geno-pheno data"
author: "Jim Holland"
date: "12/11/2019"
output: html_document
---
Data from Wisser et al 2019. Genetics
https://www.genetics.org/content/213/4/1479

34k SNP array markers on ~381 plants sampled from different generations of selection for early flowering in a tropical maize population. 

Get the marker data for Hallauer's Tuson.
https://datadryad.org/stash/dataset/doi:10.5061/dryad.q573n5tdt

From the read.me:
Quality control genotype data (45,718 variant sites) for samples from Hallauer's TusÃ³n. Tab delimited file. A header row is included. The first column ("label") lists sample names, the second column ("popdata") lists the generation to which each sample belongs, and the remaining columns correspond to the genotype data. Unphased diploid genotype calls are recorded in the following format: 1/1. Variant encoding: 1=A, 2=C, 3=G, 4=T, 5=deletion, 6=insertion (5 and 6 are used for the ZmCCT10_CACTA locus). Missing genotype calls are encoded as NA.

Data preview: 

label    popdata    PZE.101000060    PZE.101000088
C0_062314_002    0    2/1    NA
C0_062314_003    0    2/2    3/3
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(adegenet)
path = "Q:/My Drive/GGI_workshop/"
geno = read_tsv(file.path(path, "doi_10.5061_dryad.q573n5tdt__v2", "HT381_QC47518_gtype.txt.gz"))
```

What does this data set look like?
```{r}
#str(geno)
geno[1:3,1:6]
```
Notice the first column is a label for the individual, the second column indicates which generation of selection (sub-population) it belongs to.

Let's do some serious filtering against missing data, dropping lots of markers, mainly because for demonstration I want a smaller marker data set.
```{r}
missing = colSums(is.na(geno[,-c(1:2)]))/nrow(geno)
hist(missing)
```
```{r}
keep.columns = missing == 0
sum(keep.columns)
geno2 = geno[, c(F, F, keep.columns)]
dim(geno2)
```

Numericalize the genotypes to minor allele counts, 0, 1, 2 into a genlight object. Here is a hacky DIY way to do it. This will fail if there are more than 3 genotypic classes per locus, so first make a helper function to identify loci with > 3 classes, then we will drop those before numericalizing.
```{r}
classes3 = function(x){
  length(table(x)) == 3
}

keep.columns2 = apply(geno2, 2, classes3)
geno3 = geno2[keep.columns2]
dim(geno3)
```
Now create and apply a function to numericalize the genotype calls
```{r}
numericalize = function(x){
  geno.freqs = as.matrix(table(x))
  labels = row.names(geno.freqs)
  split.labels = strsplit(labels, "/", fixed = T)
  which.het = sapply(split.labels, FUN = function(x) x[1] != x[2])
  het.label = labels[which.het]
  minor.label = labels[which.min(geno.freqs)]
  major.label = labels[which.max(geno.freqs)]
  translator = c(0, 1, 2)
  names(translator) = c(major.label, het.label, minor.label)
  new.x = as.numeric(translator[x])
}
```

Try the function on the first two columns
```{r}
checkit = apply(geno3[,1:2], 2, FUN = numericalize)
checkit[1:20,]
```
Compare to original data
```{r}
geno3[1:20,1:2]
```
Now apply the function to all of the columns
```{r}
geno.num = apply(geno3[,], 2, FUN = numericalize)
str(geno.num)
```
There are some problems with my conversions, however:
```{r}
geno.num[1:10, 6]
```

```{r}
geno3[1:10, 6]
```

```{r}
table(geno3[,6])
```
You can see the problem is an excess of hets at this marker. That messes up my function. The better way is to write a function that handles edge cases, but for now let's just get rid of those problematic markers:
```{r}
ok.markers = colSums(is.na(geno.num)) == 0
sum(ok.markers)
```
```{r}
geno.num = geno.num[,ok.markers]
dim(geno.num)
```

Notice some nice properties of this numericalized data matrix:
1. Since the genotypic scores are now COUNTS of minor alleles in each individual, the minor allele frequencies are just the mean of the counts divided by two:
```{r}
maf = colMeans(geno.num)/2
hist(maf)
```
We can also compute minor allele frequencies by each generation separately and use that information to find markers that have changed allele frequency over generations:
```{r}
generations = geno[,2]
maf0 = colMeans(geno.num[generations == 0,])/2
maf10 = colMeans(geno.num[generations == 10,])/2
maf.dif = maf10 - maf0
hist(maf0)
```
```{r}
hist(maf10)
```
```{r}
hist(maf.dif)
```

2. The correlations of the columns are square roots of Linkage Disequilibrium r^2 values. So, if we get the correlation matrix for the columns and square those values, we have the LD r^2 matrix for the markers. For example, here are the correlations and LD r^2 values for the first 100 markers:
```{r}
r.mat = cor(geno.num[,1:100])
r2.mat = r.mat^2
heatmap(r2.mat, Rowv = NA, Colv = NA)
```
These markers are ordered by position (pretty sure!), you can see the small blocks of LD around the diagonal indicating local LD.

3. The correlations among individuals represent genomic relationships. Individuals with more similar marker profiles have higher correlation indicating they share more alleles:
```{r}
ind.cor = cor(t(geno.num))
dim(ind.cor)
```
The individuals are sorted by generation, so if allele frequencies changed over generations, we expect to see a little bit of higher relationship within than among generations:
```{r}
heatmap(ind.cor, Rowv = NA, Colv = NA)
```
It's pretty limited, but you can see a little bit of correlation among individuals in the blocks by generation (the individuals are ordered by generation).

4. The covariances among the individuals, scaled by the heterozygosity of the markers is the so-called VanRaden relationship matrix and it is scaled to the additive genetic covariances among the individuals. We need to center the markers first.
```{r}
Z = scale(geno.num, center = T, scale = F)
ZZpr = Z%*%t(Z)
denom = 2*sum(maf*(1 - maf))
K = ZZpr/denom
K[1:5, 1:5]
```
Diagonal elements of this matrix estimate 1 + F where F is the genomic inbreeding coefficient for each individual
```{r}
F.vals = diag(K) - 1
hist(F.vals)
```

We can check how inbreeding has changed over generations:
```{r}
df.F = data.frame(generations, F.vals)
names(df.F)[1] = "Gen"
df.F %>% group_by(Gen) %>%
  summarize(MeanF = mean(F.vals))
```
Recall that according to breeding records, the individuals were derived from random mating within generations, so we expect something close to HWE (with F = 0) within generations. When you pool data across multiple HWE populations with different allele frequencies, you will observe F > 0 in the meta-population due to differentiation (Fst). As an example, compare what happens when we estimate the relationship matrix ONLY within the initial generation 10 individuals:
```{r}
geno10 = geno.num[generations == 10,]
Z10 = scale(geno10, center = T, scale = F)
ZZpr10 = Z10%*%t(Z10)
maf10 = colMeans(geno10)/2
denom10 = 2*sum(maf10*(1 - maf10))
K10 = ZZpr10/denom10
K10[1:5, 1:5]
```
```{r}
mean(diag(K10)) - 1
```
Compare this value very close to zero based ONLY on generation 10 individuals to the meta-population mean inbreeding estimate near 0.10 for these same samples.
